### PR TITLE
Merged Whitelist functionality with src/scrape.py. Now checks for pre…

### PR DESCRIPTION
**Pull request by bonham79**
…sence of whitelist and writes separate tsv as {original file name}_filtered.tsv. Update generate_summary to reflect if file is filtered through a whitelist. CHANGELOG and README update accordingly.

- [x] Updated `Unreleased` in `CHANGELOG.md` to reflect the changes in code or data.  #155

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -29,9 +29,13 @@ Unreleased
...
 -   Only allowing Latin spellings in Maltese lexicon. (\#166).
 -   Split `mon` into Cyrillic and Mongol Bichig (\#179).
--   Added Vietnamese extraction function. (\#181)
--   Modified pron selector in Latin extraction function. (\#183)
-
+-   Added Vietnamese extraction function. (\#181).
+-   Modified pron selector in Latin extraction function. (\#183).
+-   Merged whitelist.py into 'big scrape' script. src scrape.py now checks for
+    existence of whitelist file during scrape to create second 'filtered' tsv.
+    New tsv stored in tsv folder under {"default tsv filename"}_filtered.tsv.
```
Comment (kylebgorman) [inline CHANGELOG.md:36]: "TSV" in prose (it's an acronym), thus:

   The new TSV file is placed under `tsv/*_filtered.tsv`.

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,9 +29,13 @@ Unreleased
...
+    (\#154).
+-   Updated generate_summary to reflect presence of 'filtered' tsv. (\#154)
```
Comment (kylebgorman) [inline CHANGELOG.md:38]: ...filtered TSV.

File: data/wikipron/src/README.md [hunk 3]
```diff
@@ -2,13 +2,15 @@
...
 [`scrape.py`](scrape.py) calls WikiPron's scraping functions on all Wiktionary
-languages with over 100 entries. [`generate_summary.py`](generate_summary.py)
-generates a [README](../README.md) and a [TSV](../languages_summary.tsv) with
-selected information regarding the contents of the TSVs [`scrape.py`](scrape.py)
-generated and the configuration settings that were passed to scrape.
-[`postprocess`](postprocess) sorts and removes entries in each TSV if they have
-the same graphemic form and phonetic/phonemic form as a previous entry. In
-addition it splits TSVs containing multiple scripts (Arabic, Cyrillic, etc.)
+languages with over 100 entries. If a whitelist entry is present for language,
+will generate an extra 'filtered' file only containing whitelist phones/phonemes.
```
Comment (kylebgorman) [inline data/wikipron/src/README.md:6]: Not sure why 'filtered' is quoted here...the meaning is clear, and I don't think you're doing any sort of ironic distance ;)

File: data/wikipron/src/README.md [hunk 4]
```diff
@@ -2,13 +2,15 @@
 ====================
```
Comment (kylebgorman) [inline data/wikipron/src/README.md:2]: Can you reflow this file? I use pandoc for this:

    pandoc -f markdown -t markdown source tm; mv tmp source

or something like that.

File: data/wikipron/src/scrape.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
```
Comment (kylebgorman) [inline data/wikipron/src/scrape.py:1]: Is this change intentional? (For me `python` is an alias for `python3`.)

File: data/wikipron/src/scrape.py [hunk 6]
```diff
@@ -18,23 +18,57 @@
...
             try:
-                for (word, pron) in wikipron.scrape(config):
-                    print(f"{word}\t{pron}", file=source)
+                scrape_results = wikipron.scrape(config)
+                # Given whitelist, opens up a second tsv for scraping.
+                if whitelist_set:
+                    with open(tsv_path_filtered, "w") as source_filtered:
+                        for (word, pron) in scrape_results:
+                            if _filter(word, pron, whitelist_set):
+                                print(f"{word}\t{pron}", file=source_filtered)
```
Comment (kylebgorman) [inline data/wikipron/src/scrape.py:57]: This is pretty minor but rather than building this string to print twice, I would maybe do something like:

    line = f"{word}\t{pron}
    if _filter(...):
        print(line, file=...)
    print(line)

File: data/wikipron/src/scrape.py [hunk 7]
```diff
@@ -18,23 +18,57 @@
...
+        return True
+
+
 def _call_scrape(
-    lang_settings: Dict[str, str], config: wikipron.Config, tsv_path: str
+    lang_settings: Dict[str, str],
+    config: wikipron.Config,
+    tsv_path: str,
+    whitelist_set: FrozenSet[str] = None,
+    tsv_path_filtered: str = "",
```
Comment (kylebgorman) [inline data/wikipron/src/scrape.py:46]: Maybe just call this `tsv_filtered_path` instead?

File: data/wikipron/src/scrape.py [hunk 8]
```diff
@@ -45,19 +79,65 @@ def _call_scrape(
...
 
 def _build_scraping_config(
     config_settings: Dict[str, Any], wiki_name: str, dialect_suffix: str = ""
 ) -> None:
     path_affix = f'../tsv/{config_settings["key"]}_{dialect_suffix}'
+    whitelist_path_affix = (
+        f"../whitelist/{config_settings['key']}_{dialect_suffix}"
+    )
+
+    # Configures phonemic TSV
```
Comment (kylebgorman) [inline data/wikipron/src/scrape.py:96]: Period at end of comment (and below) please. (I don't have a style guide to point to this but it's how we've done things.)

**Comment by lfashby** [comment]
This is a neat way of utilizing the whitelists. 

Don't these changes require a modification of our `postprocess` script? The whitelist section of `postprocess` is now redundant as far as I can tell.
#155 indicates that we want to keep both the unfiltered and filtered TSVs. Is that still our desire? I ran these changes locally and found that `generate_summary.py` can't handle the new filtered file names. There are several options of ways to fix that problem, but even when fixed it leaves the question of how the filtered files should be labelled in the README table of the TSVs so that they can be differentiated from the regular files. Perhaps another column indicating whether or not filtering has been applied?

One of the motivations (at least from a few months) behind incorporating a whitelisting value or something similar into entries in `languages.json` was so that `generate_summary.py` could handle filtered files in a similar fashion to how dialects are handled. Perhaps we don't need to go with that approach, but we do, I think, need a way of differentiating the `eng_uk_phonemic_filtered.tsv` row from the `eng_uk_phonemic.tsv`row in the TSV table README.

**Comment by bonham79** [comment]
What outputs were you getting for filtered files? generate_summary should ideally be outputting along the lines of "phonemic_filtered" for its transcription value (not the cleanest but seemed to have least impact on the script). Is this not the case?

**Comment by lfashby** [comment]
Sorry, probably should have explained the error I got. Running `generate_summary.py` with a file like `eng_us_phonemic_filtered.tsv` in the `tsv` directory will lead to a key error on [line 31](https://github.com/kylebgorman/wikipron/blob/75339520aaf78e4b151d7794dde277ba37a266ed/data/wikipron/src/generate_summary.py#L31) because the key isolated from line 23 will be `us_phonemic`.
I overlooked the modifications to `transcription_level`. I think that should adequately differentiate the files in the README table. Maybe not the cleanest solution, but certainly the easiest, which is fine with me.

**Comment by bonham79** [comment]
Ah, I see. I was error testing on the Greek so didn't see dialect issues. Is there any longterm plan/design philosophy for the indexing to not simply look for "phone" instead of a position underscore? It seems the most reliable for file naming conventions but not sure if that was intentionally avoided.

**Comment by kylebgorman** [comment]
Let's make sure both of them (filtered and unfiltered) end up in the table, by whatever means necessary. I guess the file name parsing needs to be slightly more robust to handle dialect, filtered/not, and script.

**Comment by lfashby** [comment]
I think we went with indexing the underscores because originally all the files were just `isokey_phonetic.tsv` or `isokey_phonemic.tsv`. We probably should have changed things when more underscores were added.

One other note, I haven't tested this but I'd imagine the new filtered file names will break `generalized_split.py` as well (more [underscore indexing](https://github.com/kylebgorman/wikipron/blob/04130daf743ccd1f39d968c4b3662b0bab0a20a2/data/wikipron/src/generalized_split.py#L34)).

**Comment by bonham79** [comment]
Alright I'm just going to push the current changes for now and then start on looking at the issue cascade then.

**Comment by bonham79** [comment]
So it looks like generalized_split was robust enough already. Just needed a tweak as far as I can tell. And whitelist can be removed for redundancies. May need a more in-depth fix for generate_summary and generalized_split where postioning of file name is less important for quality of life? (Thinking just have them load language codes (dialect, filtered, ISO) and search for their regex values without concern for order. Probably unnecessary but could avoid future issues.)

File: data/wikipron/src/postprocess [hunk 9]
```diff
@@ -10,16 +10,6 @@ split() {
...
-        OUTPUT_TSV=$(mktemp)
-        # TODO: Redirect these reports somewhere sensible.
-        ./whitelist.py "${INPUT_TSV}" "${WHITELIST}" "${OUTPUT_TSV}"
-        mv "${OUTPUT_TSV}" "${INPUT_TSV}"
-    done
-}
-
 main() {
     time split
     time whitelist
```
Comment (lfashby) [inline data/wikipron/src/postprocess:15]: I think we can get rid of this.

File: data/wikipron/src/scrape.py [hunk 10]
```diff
@@ -18,46 +18,127 @@
...
         with open(tsv_path, "w") as source:
             try:
-                for (word, pron) in wikipron.scrape(config):
-                    print(f"{word}\t{pron}", file=source)
+                scrape_results = wikipron.scrape(config)
+                # Given whitelist, opens up a second tsv for scraping.
+                if whitelist_set:
+                    with open(tsv_filtered_path, "w") as source_filtered:
+                        for (word, pron) in scrape_results:
+                            line = f"{word}\t{pron}
```
Comment (lfashby) [inline data/wikipron/src/scrape.py:56]: Need a closing `"` here?

**Comment by lfashby** [comment]
Other than those two things above, looks great to me.

File: data/wikipron/src/scrape.py (hunk 10 above)
Comment (bonham79) [reply data/wikipron/src/scrape.py]: Sorry about that. My text editor is being finicky and didn't register that edit.

File: data/wikipron/src/README.md [hunk 11]
```diff
@@ -39,11 +41,12 @@ Steps used to update the dataset
     -   By default `cut_off_date` in `main()` is set using
         `datetime.date.today().isoformat()` but can be set manually using an ISO
         formatted string (ex. "2019-10-31").
-3.  Run [`postprocess`](postprocess).
+3.  Run [`postprocess`](postprocess).filtered
```
Comment (kylebgorman) [inline data/wikipron/src/README.md:44]: typo here?

File: data/wikipron/src/scrape.py [hunk 12]
```diff
@@ -18,46 +18,127 @@
...
-    # Log and remove TSVs for languages that failed
+    # Log and remove TSVs for languages that failed.
     # to be scraped within 10 retries.
     logging.info(
         'Failed to scrape "%s" within 10 retries. %s',
         lang_settings["key"],
         lang_settings,
     )
+    # Checks if second TSV was opened.
+    if os.path.exists(tsv_filtered_path):
```
Comment (kylebgorman) [inline data/wikipron/src/scrape.py:84]: Suggest using try/except here instead of testing for existence. In theory it can help you avoid a weird race condition.

File: data/wikipron/src/README.md (hunk 11 above)
Comment (bonham79) [reply data/wikipron/src/README.md:44]: Definitely. I'm never using atom again.

File: data/wikipron/src/scrape.py (hunk 12 above)
Comment (bonham79) [reply data/wikipron/src/scrape.py:84]: Could you go into more detail? I didn't know that could happen here.

Comment (kylebgorman) [reply data/wikipron/src/scrape.py:84]: There's no particular reason to think it would occur here but the file could *in theory* be materialized between line 84 and line 85, in which case it would not be removed, or it could be deleted between lines 84 and 85, in which case this would throw an uncaught exception. So the general recommendation is try and fail.

**Comment by bonham79** [comment]
> Do you also want to remove `whitelist.py` in `src/`?

I'm thinking it can still be left in for testing purposes? Just to check if the whitelists being passed are giving effective results without needing to call a scrape.

**Comment by kylebgorman** [comment]
Okay, fine with me to leave `whitelist.py` in there.
